velox/functions/remote/utils/restserver/RemoteFunctionRestService.h

```diff
@@ -0,0 +1,137 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <boost/asio.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+
+#include "velox/common/memory/Memory.h"
+#include "velox/vector/VectorStream.h"
+
+namespace facebook::velox::functions {
+
+/// @brief Manages an individual HTTP session.
+/// Handles reading HTTP requests, processing them, and sending responses.
+/// This class re-hosts Velox functions and allows testing their functionality.
+class RestSession : public std::enable_shared_from_this<RestSession> {
+ public:
+  RestSession(boost::asio::ip::tcp::socket socket);
+
+  /// Starts the session by initiating a read operation.
+  void run();
+
+ private:
+  // Initiates an asynchronous read operation.
+  void doRead();
+
+  // Called when a read operation completes.
+  void onRead(boost::beast::error_code ec, std::size_t bytes_transferred);
+
+  // Processes the HTTP request and prepares a response.
+  void handleRequest(
+      boost::beast::http::request<boost::beast::http::string_body> req);
+
+  // Called when a write operation completes.
+  void onWrite(
+      bool close,
+      boost::beast::error_code ec,
+      std::size_t bytes_transferred);
+
+  // Closes the socket connection.
+  void doClose();
+
+  // Helper to handle exceptions: logs error and sends HTTP 500
+  void handleException(const std::exception& ex);
+
+  // Helper to extract function name from /{functionName}
+  bool extractFunctionName(
+      const boost::beast::http::request<boost::beast::http::string_body>& req,
+      std::string& outFunctionName);
+
+  // Sends a response with status, contentType, and body. Then calls
+  // async_write.
+  void sendResponse(
+      boost::beast::http::status status,
+      const std::string& contentType,
+      const std::string& body,
+      bool closeConnection);
+
+  // Helper to ensure the HTTP method is POST, else sends an error response.
+  bool ensurePostMethod(
+      const boost::beast::http::request<boost::beast::http::string_body>& req);
+
+  // Helper to ensure Content-Type is either "application/X-presto-pages" or
+  // "application/X-spark-unsafe-row". Otherwise sends an error response.
+  bool ensureValidContentType(
+      const boost::beast::http::request<boost::beast::http::string_body>& req);
+
+  // Helper to ensure Accept is either "application/X-presto-pages" or
+  // "application/X-spark-unsafe-row". Otherwise sends an error response.
+  bool ensureValidAccept(
+      const boost::beast::http::request<boost::beast::http::string_body>& req);
+
+  // Sends a success response with the given payload.
+  void sendSuccessResponse(folly::IOBuf&& payload);
+
+  void handleRemoteAbs(

```

Comment:
Would be good if this code is written in a more generic way.

So define an API for a RemoteFunctionHandler. The test author will need to register their 
 remote function handlers with the RestSession. The RestSession will look up its registry (by function name) and invoke the returned function handler. 
 
That would give flexibility to write different tests with different functions instead of a service with these fixed functions.